pkg/apis/ceph.rook.io/v1/topic.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// compile-time assertions ensures CephBucketTopic implements webhook.Validator so a webhook builder
+// will be registered for the validating webhook.
+var _ webhook.Validator = &CephBucketTopic{}
+
+func ValidateHTTPSpec(s *HTTPEndpointSpec) error {
+	return nil
+}
+
+func ValidateAMQPSpec(s *AMQPEndpointSpec) error {
+	if s == nil {
+		return errors.New("missing AMQP spec")
+	}
+	return nil
+}
+
+func ValidateKafkaSpec(s *KafkaEndpointSpec) error {
+	return nil
+}
+
+// ValidateTopicSpec validate the bucket notification topic arguments
+func ValidateTopicSpec(t *CephBucketTopic) error {
+	if t.Name == "" {
+		return errors.New("missing name")
+	}
+	var urlSchema string
+	endpointURL, err := url.Parse(t.Spec.Endpoint)
+	if err != nil {
+		return err
+	}
+	if urlSchema = strings.ToLower(endpointURL.Scheme); urlSchema == "amqp" || urlSchema == "amqps" {
+		if err := ValidateAMQPSpec(t.Spec.AMQP); err != nil {
+			return err
+		}
+		if t.Spec.HTTP != nil || t.Spec.Kafka != nil {
+			return errors.New("multiple endpoint specs")
+		}
+	} else if urlSchema == "http" || urlSchema == "https" {
+		if err := ValidateHTTPSpec(t.Spec.HTTP); err != nil {
+			return err
+		}
+		if t.Spec.AMQP != nil || t.Spec.Kafka != nil {
+			return errors.New("multiple endpoint specs")
+		}
+	} else if urlSchema == "kafka" {
+		if err := ValidateKafkaSpec(t.Spec.Kafka); err != nil {
+			return err
+		}
+		if t.Spec.AMQP != nil || t.Spec.HTTP != nil {
+			return errors.New("multiple endpoint specs")
+		}
+
+	} else {
+		return errors.Errorf("invalid endpoint schema: %q", urlSchema)
+	}
+
+	return nil
+}
+
+func (t *CephBucketTopic) ValidateCreate() error {
+	logger.Infof("validate create cephbuckettopic %v", t)
+	if err := ValidateTopicSpec(t); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *CephBucketTopic) ValidateUpdate(old runtime.Object) error {
+	logger.Info("validate update cephbuckettopic")
+	if err := ValidateTopicSpec(t); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *CephBucketTopic) ValidateDelete() error {
+	logger.Info("validate delete cephbuckettopic")
+	return nil
+}
```

Comment:
Do we need different validate functions, ValidateTopicSpec() can handle all the scenarios, Or I am missing the wider picture?

Reply:
these are needed so we comply with the `webhook.Validator` interface